server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -0,0 +1,1192 @@
+    if (startDelay > WARN_DELAY) {
+      log.warn("Ingestion was throttled for [%,d] millis because persists were pending.", startDelay);
+    }
+    runExecStopwatch.stop();
+    resetNextFlush();
+
+    return future;
+  }
+
+  @Override
+  public ListenableFuture<SegmentsAndMetadata> push(
+      final List<SegmentIdentifier> identifiers,
+      final Committer committer
+  )
+  {
+    final Map<SegmentIdentifier, Sink> theSinks = Maps.newHashMap();
+    for (final SegmentIdentifier identifier : identifiers) {
+      final Sink sink = sinks.get(identifier);
+      if (sink == null) {
+        throw new NullPointerException("No sink for identifier: " + identifier);
+      }
+      theSinks.put(identifier, sink);
+      sink.finishWriting();
+    }
+
+    return Futures.transform(
+        persistAll(committer),
+        new Function<Object, SegmentsAndMetadata>()
+        {
+          @Override
+          public SegmentsAndMetadata apply(Object commitMetadata)
+          {
+            final List<DataSegment> dataSegments = Lists.newArrayList();
+
+            for (Map.Entry<SegmentIdentifier, Sink> entry : theSinks.entrySet()) {
+              if (droppingSinks.contains(entry.getKey())) {
+                log.info("Skipping push of currently-dropping sink[%s]", entry.getKey());
+                continue;
+              }
+
+              final DataSegment dataSegment = mergeAndPush(entry.getKey(), entry.getValue());
+              if (dataSegment != null) {
+                dataSegments.add(dataSegment);
+              } else {
+                log.warn("mergeAndPush[%s] returned null, skipping.", entry.getKey());
+              }
+            }
+
+            return new SegmentsAndMetadata(dataSegments, commitMetadata);
+          }
+        },
+        mergeExecutor
+    );
+  }
+
+  /**
+   * Insert a barrier into the merge-and-push queue. When this future resolves, all pending pushes will have finished.
+   * This is useful if we're going to do something that would otherwise potentially break currently in-progress
+   * pushes.
+   */
+  private ListenableFuture<?> mergeBarrier()
+  {
+    return mergeExecutor.submit(
+        new Runnable()
+        {
+          @Override
+          public void run()
+          {
+            // Do nothing
+          }
+        }
+    );
+  }
+
+  /**
+   * Merge segment, push to deep storage. Should only be used on segments that have been fully persisted. Must only
+   * be run in the single-threaded mergeExecutor.
+   *
+   * @param identifier sink identifier
+   * @param sink       sink to push
+   *
+   * @return segment descriptor, or null if the sink is no longer valid
+   */
+
+  private DataSegment mergeAndPush(final SegmentIdentifier identifier, final Sink sink)
+  {
+    // Bail out if this sink is null or otherwise not what we expect.
+    if (sinks.get(identifier) != sink) {
+      log.warn("Sink for segment[%s] no longer valid, bailing out of mergeAndPush.", identifier);
```

Comment:
perhaps a sink was dropped and then a different one with the same identifier was re-added.